internal/service/accrual_service.go

```diff
@@ -0,0 +1,151 @@
+package service
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/golangTroshin/gophermat/internal/config"
+	"github.com/golangTroshin/gophermat/internal/model"
+	"gorm.io/gorm"
+)
+
+type APIResponse struct {
+	Order   string  `json:"order"`
+	Status  string  `json:"status"`
+	Accrual float64 `json:"accrual,omitempty"`
+}
+
+var (
+	pause    bool
+	pauseMux sync.Mutex
+)
+
+func fetchOrderStatus(orderNumber string) (*APIResponse, error) {
+	url := fmt.Sprintf("%s/api/orders/%s", config.Options.AccrualSystemAddress, orderNumber)
+
+	resp, err := http.Get(url)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode == http.StatusNoContent {
+		return nil, fmt.Errorf("order not found in the accrual system")
+	} else if resp.StatusCode == http.StatusTooManyRequests {
+		retryAfter := resp.Header.Get("Retry-After")
+		if retryAfter != "" {
+			retryAfterSeconds, err := strconv.Atoi(retryAfter)
+			if err != nil {
+				return nil, fmt.Errorf("error parsing Retry-After header: %v", err)
+			}
+
+			pauseMux.Lock()
+			pause = true
+			pauseMux.Unlock()
+
+			log.Printf("Too many requests, pausing for %d seconds", retryAfterSeconds)
+			time.Sleep(time.Duration(retryAfterSeconds) * time.Second)
+
+			pauseMux.Lock()
+			pause = false
+			pauseMux.Unlock()
+
+			return nil, fmt.Errorf("retry after pause due to 429")
+		}
+		return nil, fmt.Errorf("too many requests, retry later")
+	} else if resp.StatusCode == http.StatusInternalServerError {
+		return nil, fmt.Errorf("internal server error")
+	}
+
+	var apiResponse APIResponse
+	err = json.NewDecoder(resp.Body).Decode(&apiResponse)
+	if err != nil {
+		return nil, err
+	}
+
+	return &apiResponse, nil
+}
+
+func updateOrderStatus(db *gorm.DB, order *model.Order, apiResponse *APIResponse) error {
+	newStatus := apiResponse.Status
+	if apiResponse.Status == "REGISTERED" {
+		newStatus = "PROCESSING"
+	}
+
+	order.Status = newStatus
+	if apiResponse.Status == "PROCESSED" {
+		order.Accrual = apiResponse.Accrual
+
+		var user model.User
+		if err := db.Preload("Balance").First(&user, order.UserID).Error; err != nil {
+			return err
+		}
+
+		user.Balance.Current += apiResponse.Accrual
+
+		if err := db.Save(&user.Balance).Error; err != nil {
+			return err
+		}
+	}
+
+	return db.Save(order).Error
+}
+
+func processOrders(db *gorm.DB) {
+	var orders []model.Order
+	db.Where("status IN ?", []string{"NEW", "PROCESSING"}).Find(&orders)
+
+	for _, order := range orders {
+		pauseMux.Lock()
+		paused := pause
+		pauseMux.Unlock()
+
+		if paused {
+			log.Println("Waiting for pause to be lifted...")
+			for paused {
+				time.Sleep(1 * time.Second)
+				pauseMux.Lock()
+				paused = pause
+				pauseMux.Unlock()
+			}
+		}
```

Comment:
В случае с RWLock, мы все еще можем читать, но записывать в этот момент никто не сможет. Таким образом другие читатели не блокируются.

```suggestion
        pauseMux.RLock()
	paused := pause
	pauseMux.RUnlock()

	if paused {
		log.Println("Waiting for pause to be lifted...")
		for paused {
			time.Sleep(1 * time.Second)
			pauseMux.Lock()
			paused = pause
			pauseMux.RUnlock()
		}
	}
```